File: d2c.py
from __future__ import annotations

import time
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Literal, Optional, Self

import requests
from pydantic import BaseModel, model_validator

SandboxType = Literal["postgres", "redis"]


class SandboxResponse(BaseModel):
    id: str
    sandbox_type: SandboxType
    status: Optional[str]
    created_at: datetime
    credentials: Optional[dict[str, Any]] = None
    url: Optional[str] = None

    @model_validator(mode="after")
    def validate_credentials(self) -> Self:
        try:
            if self.sandbox_type == "postgres":
                self.url = f"postgresql://{self.credentials['user']}:{self.credentials['password']}@{self.credentials['host']}:{self.credentials['port']}/{self.credentials['database']}"
            elif self.sandbox_type == "redis":
                self.url = f"redis://{self.credentials['user']}:{self.credentials['password']}@{self.credentials['host']}:{self.credentials['port']}"
        except Exception:
            pass
        return self


class Dev2CloudError(Exception):
    """Raised when the API returns an error response."""

    def __init__(self, status_code: int, detail: str) -> None:
        self.status_code = status_code
        self.detail = detail
        super().__init__(f"[{status_code}] {detail}")


class Dev2Cloud:
    """Client for the Dev2Cloud sandbox management API.

    Example::

        from d2c import Dev2Cloud

        client = Dev2Cloud(api_key="your-api-key")

        # Create a sandbox and wait until it's running
        sandbox = client.create_sandbox("postgres")
        print(sandbox.credentials)

        # List all active sandboxes
        for sb in client.list_sandboxes():
            print(sb.id, sb.status)

        # Clean up
        client.delete_sandbox(sandbox.id)
    """

    def __init__(self, api_key: str, api_url: str = "https://api.dev2.cloud") -> None:
        """Initialise the Dev2Cloud client.

        Args:
            api_key: API key used to authenticate requests (sent as ``X-Api-Key`` header).
            api_url: Base URL of the Dev2Cloud API. Defaults to ``https://api.dev2.cloud``.
        """
        self._api_url = api_url.rstrip("/")
        self._session = requests.Session()
        self._session.headers.update({"X-Api-Key": api_key})

    # -- helpers ----------------------------------------------------------

    def _url(self, path: str) -> str:
        return f"{self._api_url}{path}"

    @staticmethod
    def _raise_on_error(response: requests.Response) -> None:
        if response.ok:
            return
        try:
            detail = response.json().get("detail", response.text)
        except ValueError:
            detail = response.text
        raise Dev2CloudError(response.status_code, detail)

    @staticmethod
    def _parse_sandbox(data: dict[str, Any]) -> SandboxResponse:
        return SandboxResponse(
            id=data["id"],
            sandbox_type=data["sandbox_type"],
            status=data.get("status"),
            created_at=datetime.fromisoformat(data["created_at"]),
            credentials=data.get("credentials"),
        )

    # -- public API -------------------------------------------------------

    def list_sandboxes(self) -> list[SandboxResponse]:
        """Lists all active sandboxes for the authenticated user.

        Returns:
            A list of sandbox objects sorted by creation time.

        Raises:
            Dev2CloudError: If the API returns an error response.
        """
        response = self._session.get(self._url("/api/v1/sandboxes"))
        self._raise_on_error(response)
        return [self._parse_sandbox(item) for item in response.json()]

    def create_sandbox(
        self, sandbox_type: SandboxType, *, timeout: float = 180
    ) -> SandboxResponse:
        """Creates a new sandbox and waits for it to be ready.

        Provisions a sandbox and polls its status once per second until it
        transitions to ``running`` (credentials will be available) or
        ``failed``.

        Args:
            sandbox_type: The type of sandbox to create (``"postgres"`` or ``"redis"``).
            timeout: Maximum seconds to wait for the sandbox to become
                ready. Defaults to 180 (3 minutes).

        Returns:
            The sandbox object with ``running`` status and connection credentials.

        Raises:
            Dev2CloudError: If the sandbox transitions to ``failed`` or
                does not become ready within ``timeout`` seconds.
        """
        response = self._session.post(
            self._url("/api/v1/sandboxes"),
            json={"sandbox_type": sandbox_type},
        )
        self._raise_on_error(response)
        sandbox_id = response.json()["id"]

        deadline = time.monotonic() + timeout
        while True:
            if time.monotonic() >= deadline:
                raise Dev2CloudError(
                    0, f"Sandbox {sandbox_id} did not become ready within {timeout}s"
                )
            time.sleep(1)
            sandbox = self.get_sandbox(sandbox_id)
            if sandbox.status != "pending":
                break

        if sandbox.status == "failed":
            raise Dev2CloudError(0, f"Sandbox {sandbox_id} failed to provision")

        return sandbox

    def get_sandbox(self, sandbox_id: str) -> SandboxResponse:
        """Gets a sandbox by its ID.

        Args:
            sandbox_id: The unique identifier of the sandbox.

        Returns:
            The sandbox object including its current status and credentials.

        Raises:
            Dev2CloudError: If the API returns an error response.
        """
        response = self._session.get(self._url(f"/api/v1/sandboxes/{sandbox_id}"))
        self._raise_on_error(response)
        return self._parse_sandbox(response.json())

    def delete_sandbox(self, sandbox_id: str) -> None:
        """Permanently deletes a sandbox.

        This action is irreversible. Connection credentials are revoked
        immediately.

        Args:
            sandbox_id: The unique identifier of the sandbox to delete.

        Raises:
            Dev2CloudError: If the API returns an error response.
        """
        response = self._session.delete(self._url(f"/api/v1/sandboxes/{sandbox_id}"))
        self._raise_on_error(response)

    def delete_all(self) -> list[str]:
        """Deletes all active sandboxes.

        Fetches the current sandbox list and deletes each one. Deletion
        errors for individual sandboxes are silently ignored so that one
        failure does not prevent the remaining sandboxes from being removed.

        Returns:
            A list of sandbox IDs that were successfully deleted.
        """
        sandboxes = self.list_sandboxes()
        deleted: list[str] = []
        for sb in sandboxes:
            try:
                self.delete_sandbox(sb.id)
                deleted.append(sb.id)
            except Dev2CloudError:
                pass
        return deleted

File: d2c.js
/**
 * @typedef {"postgres"|"redis"} SandboxType
 */

/**
 * @typedef {Object} SandboxResponse
 * @property {string} id
 * @property {SandboxType} sandboxType
 * @property {string|null} status
 * @property {Date} createdAt
 * @property {Object.<string, any>|null} [credentials]
 */

class Dev2CloudError extends Error {
  /**
   * Raised when the API returns an error response.
   * @param {number} statusCode
   * @param {string} detail
   */
  constructor(statusCode, detail) {
    super(`[${statusCode}] ${detail}`);
    this.name = "Dev2CloudError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
}

class Dev2Cloud {
  /**
   * Client for the Dev2Cloud sandbox management API.
   *
   * @example
   * const client = new Dev2Cloud("your-api-key");
   *
   * // Create a sandbox and wait until it's running
   * const sandbox = await client.createSandbox("postgres");
   * console.log(sandbox.credentials);
   *
   * // List all active sandboxes
   * for (const sb of await client.listSandboxes()) {
   *   console.log(sb.id, sb.status);
   * }
   *
   * // Clean up
   * await client.deleteSandbox(sandbox.id);
   */

  /**
   * Initialise the Dev2Cloud client.
   *
   * @param {string} apiKey - API key used to authenticate requests (sent as `X-Api-Key` header).
   * @param {string} [apiUrl="https://api.dev2.cloud"] - Base URL of the Dev2Cloud API.
   */
  constructor(apiKey, apiUrl = "https://api.dev2.cloud") {
    this._apiUrl = apiUrl.replace(/\/+$/, "");
    this._apiKey = apiKey;
  }

  // -- helpers ----------------------------------------------------------

  /**
   * @param {string} path
   * @returns {string}
   */
  _url(path) {
    return `${this._apiUrl}${path}`;
  }

  /**
   * @param {Response} response
   * @throws {Dev2CloudError}
   */
  static async _raiseOnError(response) {
    if (response.ok) return;
    let detail;
    try {
      const body = await response.json();
      detail = body.detail ?? response.statusText;
    } catch {
      detail = response.statusText;
    }
    throw new Dev2CloudError(response.status, detail);
  }

  /**
   * @param {Object.<string, any>} data
   * @returns {SandboxResponse}
   */
  static _parseSandbox(data) {
    return {
      id: data.id,
      sandboxType: data.sandbox_type,
      status: data.status ?? null,
      createdAt: new Date(data.created_at),
      credentials: data.credentials ?? null,
    };
  }

  /**
   * @param {string} path
   * @param {RequestInit} [options]
   * @returns {Promise<Response>}
   */
  _fetch(path, options = {}) {
    const headers = {
      "X-Api-Key": this._apiKey,
      ...options.headers,
    };
    return fetch(this._url(path), { ...options, headers });
  }

  // -- public API -------------------------------------------------------

  /**
   * Lists all active sandboxes for the authenticated user.
   *
   * @returns {Promise<SandboxResponse[]>} A list of sandbox objects sorted by creation time.
   * @throws {Dev2CloudError} If the API returns an error response.
   */
  async listSandboxes() {
    const response = await this._fetch("/api/v1/sandboxes");
    await Dev2Cloud._raiseOnError(response);
    const data = await response.json();
    return data.map(Dev2Cloud._parseSandbox);
  }

  /**
   * Creates a new sandbox and waits for it to be ready.
   *
   * Provisions a sandbox and polls its status once per second until it
   * transitions to `running` (credentials will be available) or `failed`.
   *
   * @param {SandboxType} sandboxType - The type of sandbox to create (`"postgres"` or `"redis"`).
   * @param {number} [timeout=180] - Maximum seconds to wait for the sandbox to become ready.
   * @returns {Promise<SandboxResponse>} The sandbox object with `running` status and connection credentials.
   * @throws {Dev2CloudError} If the sandbox transitions to `failed` or does not become ready within `timeout` seconds.
   */
  async createSandbox(sandboxType, timeout = 180) {
    const response = await this._fetch("/api/v1/sandboxes", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sandbox_type: sandboxType }),
    });
    await Dev2Cloud._raiseOnError(response);
    let sandbox = Dev2Cloud._parseSandbox(await response.json());

    const deadline = Date.now() + timeout * 1000;
    while (sandbox.status === "pending") {
      if (Date.now() >= deadline) {
        throw new Dev2CloudError(
          0,
          `Sandbox ${sandbox.id} did not become ready within ${timeout}s`
        );
      }
      await new Promise((resolve) => setTimeout(resolve, 1000));
      sandbox = await this.getSandbox(sandbox.id);
    }

    if (sandbox.status === "failed") {
      throw new Dev2CloudError(0, `Sandbox ${sandbox.id} failed to provision`);
    }

    return sandbox;
  }

  /**
   * Gets a sandbox by its ID.
   *
   * @param {string} sandboxId - The unique identifier of the sandbox.
   * @returns {Promise<SandboxResponse>} The sandbox object including its current status and credentials.
   * @throws {Dev2CloudError} If the API returns an error response.
   */
  async getSandbox(sandboxId) {
    const response = await this._fetch(`/api/v1/sandboxes/${sandboxId}`);
    await Dev2Cloud._raiseOnError(response);
    return Dev2Cloud._parseSandbox(await response.json());
  }

  /**
   * Permanently deletes a sandbox.
   *
   * This action is irreversible. Connection credentials are revoked immediately.
   *
   * @param {string} sandboxId - The unique identifier of the sandbox to delete.
   * @throws {Dev2CloudError} If the API returns an error response.
   */
  async deleteSandbox(sandboxId) {
    const response = await this._fetch(`/api/v1/sandboxes/${sandboxId}`, {
      method: "DELETE",
    });
    await Dev2Cloud._raiseOnError(response);
  }

  /**
   * Deletes all active sandboxes.
   *
   * Fetches the current sandbox list and deletes each one. Deletion
   * errors for individual sandboxes are silently ignored so that one
   * failure does not prevent the remaining sandboxes from being removed.
   *
   * @returns {Promise<string[]>} A list of sandbox IDs that were successfully deleted.
   */
  async deleteAll() {
    const sandboxes = await this.listSandboxes();
    const deleted = [];
    for (const sb of sandboxes) {
      try {
        await this.deleteSandbox(sb.id);
        deleted.push(sb.id);
      } catch (err) {
        if (!(err instanceof Dev2CloudError)) throw err;
      }
    }
    return deleted;
  }
}

module.exports = { Dev2Cloud, Dev2CloudError };
